### PR TITLE
Add API to make `unnest` consistent with DuckDB/ClickHouse, add option for preserve_nulls, update docs

### PR DESCRIPTION
**Pull request by alamb**
This is a follow on to https://github.com/apache/arrow-datafusion/pull/7088, my third attempt to document DataFusion's `Unnest` behavior

# Which issue does this PR close?

re https://github.com/apache/arrow-datafusion/issues/7087

closes https://github.com/apache/arrow-datafusion/pull/7044

# Rationale for this change

The unnest functionality is both under documented in DataFusion as well as inconsistent with respect to null handling with DuckDB and ClickHouse. 

# What changes are included in this PR?

This implements the proposal and various feedback from @jayzhan211 @izveigor @vincev  and @smiklos from https://github.com/apache/arrow-datafusion/pull/7044#discussion_r1271437027 and elsewhere

1. Document what `unnest` does, with an example
2. Add `UnnestOptions` that controls the behavior of `unnest` with null inputs and `unest_with_options` and related functions
3. Add tests showing how to use `UnnestOptions` and 

note this PR *DOES NOT* actually change the `unnest` behavior -- I believe that should be done as a separate PR

# Are these changes tested?

Yes

# Are there any user-facing changes?
New API (but existing APIs stay the same)

### CODE REVIEW

File: datafusion/core/src/dataframe.rs [hunk 1]
```diff
@@ -151,20 +151,38 @@ impl DataFrame {
...
     pub fn unnest_column(self, column: &str) -> Result<DataFrame> {
+        self.unnest_column_with_options(column, UnnestOptions::new())
+    }
+
+    /// Expand each list element of a column to multiple rows, with
+    /// behavior controlled by [`UnnestOptions`].
+    ///
+    /// Please see the documentation on [`UnnestOptions`] for more
+    /// details about the meaning of unnest.
+    pub fn unnest_column_with_options(
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:179]: Here is the key -- add a `_with_options` variant and link to documentation

File: datafusion/core/tests/dataframe/mod.rs [hunk 2]
```diff
@@ -1044,6 +1044,82 @@ async fn unnest_columns() -> Result<()> {
...
+    let results = table_with_lists_and_nulls()
+        .await?
+        .clone()
+        .unnest_column_with_options("list", options)?
+        .collect()
+        .await;
+
+    assert_eq!(
+        results.unwrap_err().to_string(),
+        "This feature is not implemented: Unest with preserve_nulls=true"
```
Comment (alamb) [inline datafusion/core/tests/dataframe/mod.rs]: preserve_nulls is not yet supported, but here is a space for it

File: datafusion/core/src/dataframe.rs [hunk 3]
```diff
@@ -151,20 +151,38 @@ impl DataFrame {
...
+    ///
     /// ```
     /// # use datafusion::prelude::*;
     /// # use datafusion::error::Result;
     /// # #[tokio::main]
     /// # async fn main() -> Result<()> {
     /// let ctx = SessionContext::new();
     /// let df = ctx.read_csv("tests/data/example.csv", CsvReadOptions::new()).await?;
-    /// let df = df.unnest_column("a")?;
+    /// let df = df.unnest_column_wtith_optons("a")?;
```
Comment (smiklos) [inline datafusion/core/src/dataframe.rs:166]: typo

**Review by smiklos — COMMENTED** [review]
This looks great! 

Wouldn't adding new flags to the options be also breaking change vs just expanding unnest_with_options with new arguments?

File: datafusion/core/src/physical_plan/unnest.rs [hunk 4]
```diff
@@ -133,6 +146,12 @@ impl ExecutionPlan for UnnestExec {
     ) -> Result<SendableRecordBatchStream> {
         let input = self.input.execute(partition, context)?;
 
+        if self.options.preserve_nulls {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/physical_plan/unnest.rs:149]: I got this backwards -- I will fix

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/physical_plan/unnest.rs:149]: Fixed in 5b975d256

**Review by smiklos — APPROVED** [review]
This looks 🏅 


#7088 should not be marked as #closes just yet.

**Comment by alamb** [comment]
> https://github.com/apache/arrow-datafusion/pull/7088 should not be marked as #closes just yet.

Thanks, removed

**Comment by alamb** [comment]
@jayzhan211  and @izveigor  -- what do you think about merging this PR / improved API for unnest?

**Review by yjshen — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again everyone for the reviews -- I can't wait to see what we build with this

**Pull request merged**